aimmo-game/service.py

```diff
@@ -3,104 +3,51 @@
 eventlet.monkey_patch()
 
 import flask
-from flask_socketio import SocketIO, emit
+from flask_socketio import SocketIO
 
 from simulation.turn_manager import state_provider
 from simulation import map_generator
 from simulation.avatar.avatar_manager import AvatarManager
 from simulation.turn_manager import ConcurrentTurnManager
 from simulation.worker_manager import WORKER_MANAGERS
+from simulation.world_state import WorldState
 
 app = flask.Flask(__name__)
 socketio = SocketIO()
 
 worker_manager = None
 
-
-def to_cell_type(cell):
-    if not cell.habitable:
-        return 1
-    if cell.generates_score:
-        return 2
-    return 0
-
-
-def player_dict(avatar):
-    # TODO: implement better colour functionality: will eventually fall off end of numbers
-    colour = "#%06x" % (avatar.player_id * 4999)
-    return {
-        'id': avatar.player_id,
-        'x': avatar.location.x,
-        'y': avatar.location.y,
-        'health': avatar.health,
-        'score': avatar.score,
-        'rotation': 0,
-        "colours": {
-            "bodyStroke": "#0ff",
-            "bodyFill": colour,
-            "eyeStroke": "#aff",
-            "eyeFill": "#eff",
-        }
-    }
-
-
-def get_world_state():
-    with state_provider as game_state:
-        world = game_state.world_map
-        player_data = {p.player_id: player_dict(p) for p in game_state.avatar_manager.avatars}
-        grid_dict = defaultdict(dict)
-        for cell in world.all_cells():
-            grid_dict[cell.location.x][cell.location.y] = to_cell_type(cell)
-        pickups = []
-        for cell in world.pickup_cells():
-            pickup = cell.pickup.serialise()
-            pickup['location'] = (cell.location.x, cell.location.y)
-            pickups.append(pickup)
-        return {
-                'players': player_data,
-                'score_locations': [(cell.location.x, cell.location.y) for cell in world.score_cells()],
-                'pickups': pickups,
-                # TODO: experiment with only sending deltas (not if not required)
-                'map_changed': True,
-                'width': world.num_cols,
-                'height': world.num_rows,
-                'minX': world.min_x(),
-                'minY': world.min_y(),
-                'maxX': world.max_x(),
-                'maxY': world.max_y(),
-                'layout': grid_dict,
-            }
-
+# The world state know about the maps and the avatars and basically
+# handles the updates.
+world_state = WorldState(state_provider)
 
 @socketio.on('connect')
-def world_update_on_connect():
-    emit(
-        'world-update',
-        get_world_state(),
-    )
+def world_init():
+    socketio.emit('world-init')
 
+@socketio.on('client-ready')
+def client_ready(client_id):
+    world_state.ready_to_update = True
 
 def send_world_update():
-    socketio.emit(
-        'world-update',
-        get_world_state(),
-        broadcast=True,
-    )
-
+    if world_state.ready_to_update:
```

Comment:
Wait, I understand the confusion. In the 'unity' branch every client has its own world_state. That makes sense with my previous comment. Here we only have one world_state for all of them, in which case the ready_to_update is pointless. We'll remove it from this pull request since it's not strictly related to the API refactoring and add it back later.